src/loader/resolvable-promise.js

```diff
@@ -0,0 +1,84 @@
+/**
+ * A Promise with resolve() and reject() methods so that it can be fulfilled
+ * asynchronously from outside instead of the callback.
+ *
+ * Because Promise cannot be extended in the usual way, we need to do some
+ * magic to define this class.
+ */
+export default class ResolvablePromise {
+	/**
+	 * Don't construct ResolvablePromise objects directly: rely on
+	 * ResolvablePromise.new() instead.
+	 */
+	constructor() {
+		throw new Error(
+			'Don\'t construct ResolvablePromise objects directly: ' +
+				'rely on ResolvablePromise.new() instead'
+		);
+	}
+}
+
+ResolvablePromise.new = () => {
+	const capture = {};
+
+	const promise = new Promise((resolve, reject) => {
+		capture._resolve = resolve;
+		capture._reject = reject;
+	});
+
+	Object.assign(promise, capture, {
+		fulfilled: false,
+		resolved: false,
+		rejected: false,
+		resolution: undefined,
+		rejection: undefined,
+	});
+
+	promise.resolve = value => resolve(promise, value);
+	promise.reject = error => reject(promise, error);
+
+	// This is to avoid UnhandledPromiseRejectionWarning errors in node
+	promise.catch(() => {});
```

Comment:
I second this.

Reply:
Mmmm, not really because of the way the ResolvablePromise is used. I tested it when I added the line.

Having that catch does not prevent other catches from firing. If you want to see it with your eyes paste the following code in the browser's console:

```
capture = {};

a = new Promise((resolve, reject) => {
	capture.reject = reject;
})

a.catch(e => console.log('1>',e))
a.catch(e => console.log('2>',e))
a.catch(e => console.log('3>',e))

capture.reject('hola');
```

or Node and you will see that the three catches are fired. If you move the reject call before the catches it works too.

However, if you use a regular promise with several chained catches, only the first one fires but that's because catch itself returns another promise different from the original one (I think, I have not analyzed it deeply).

Anyway, given that the catch is only needed for tests I will wrap it inside an if so that we all can sleep better :-D.